src/geogenalg/analyze.py

```diff
@@ -0,0 +1,58 @@
+#  Copyright (c) 2025 National Land Survey of Finland (Maanmittauslaitos)
+#
+#  This file is part of geogen-algorithms.
+#
+#  This source code is licensed under the MIT license found in the
+#  LICENSE file in the root directory of this source tree.
+
+from geopandas import GeoDataFrame, overlay
+
+
+def calculate_coverage(
+    overlay_features: GeoDataFrame,
+    base_features: GeoDataFrame,
+    coverage_attribute: str = "coverage",
+) -> GeoDataFrame:
+    """Calculate the percentage area coverage of base features by overlay features.
+
+    Returns:
+        GeoDataFrame: Base features with new 'coverage' column (% overlap).
+
+    Raises:
+        ValueError: If CRS is missing or not projected.
+
+    """
+    if (
+        overlay_features.crs is None
+        or base_features.crs is None
+        or not overlay_features.crs.is_projected
+        or not base_features.crs.is_projected
+    ):
+        error_msg = "Both layers must have a projected CRS (metres)."
+        raise ValueError(error_msg)
+
+    overlay_features = overlay_features.copy()
+    base_features = base_features.copy()
+    base_features["base_area"] = base_features.geometry.area
+
+    base_features["base_feature_id"] = base_features.index
+
+    intersections = overlay(overlay_features, base_features, how="intersection")
+    intersections["intersect_area"] = intersections.geometry.area
+
+    combined_overlay_area = (
+        intersections.groupby("base_feature_id")["intersect_area"]
+        .sum()
+        .reset_index()
+        .rename(columns={"intersect_area": "overlay_area"})
+    )
+
+    base_features = base_features.merge(
+        combined_overlay_area, on="base_feature_id", how="left"
+    )
+    base_features["overlay_area"] = base_features["overlay_area"].fillna(0)
+    base_features[coverage_attribute] = (
+        100 * base_features["overlay_area"] / base_features["base_area"]
+    )
```

Comment:
Extra attributes (base_feature_id, overlay_area, base_area) should probably be dropped after merge as they are not needed anymore and extra attributes increase the size of the dataframe. Also, the docstring only mentions that coverage attribute is added so that's what the caller of the function expects.